kuma/core/middleware.py

```diff
@@ -27,15 +27,25 @@ def process_request(self, request):
         prefixer = Prefixer(request)
         set_url_prefixer(prefixer)
         full_path = prefixer.fix(prefixer.shortened_path)
+        lang = request.GET.get('lang')
 
-        if 'lang' in request.GET:
+        if lang in dict(settings.LANGUAGES):
             # Blank out the locale so that we can set a new one. Remove lang
             # from the query params so we don't have an infinite loop.
             prefixer.locale = ''
             new_path = prefixer.fix(prefixer.shortened_path)
-            query = dict((smart_str(k), v) for
-                         k, v in request.GET.iteritems() if k != 'lang')
-            return HttpResponsePermanentRedirect(urlparams(new_path, **query))
+            query = dict((smart_str(k), v) for k, v in request.GET.iteritems() if k != 'lang')
+
+            # Never use HttpResponsePermanentRedirect here.
+            # Its a temporary redirect and should return with http 302, not 301
+            response = HttpResponseRedirect(urlparams(new_path, **query))
+
+            # 'lang' is only used on the language selection dropdown. If this is
+            # present it is safe to set language preference as cookie for the current
+            # user.
+            response.set_cookie(settings.LANGUAGE_COOKIE_NAME, lang)
```

Comment:
We want to avoid search engine scrapers getting a language preference cookie. The use of the language switcher should prompt the user if they want us to remember their language setting. A POST is needed to create the cookie. I don't believe a middleware is the right solution for setting the cookie.